### ISSUE TITLE
ValueError ... is not in list

### ISSUE THREAD

**Issue by pseudotensor** [gradio-app/gradio#4317]
### Describe the bug

Radio button by index fails to work properly, fails.

By value works fine.  However, in my case I need to choose by index since radio text itself is not easy to use in my case.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

Run below code.  See how selecting any radio works fine, index shows up in text box.  After adding "Z" by entering into textbox and clicking ADD, then adds new entry.  However, once select Z, get failure as if Z is not in the list.

```
import gradio as gr


def change_radio(radio_state1, text):
    radio_state1.append(text)
    return radio_state1, gr.Radio.update(choices=radio_state1), ''


def change_text(choice):
    return choice


with gr.Blocks() as demo:
    choices0 = ["A", "B", "C"]
    radio_state = gr.State(choices0)
    radio = gr.Radio(
        choices0, label="Choose", type='index',
    )
    text = gr.Textbox(interactive=True)
    add = gr.Button("Add")

    radio.input(fn=change_text, inputs=radio, outputs=text)

    add.click(change_radio, inputs=[radio_state, text], outputs=[radio_state, radio, text])

if __name__ == "__main__":
    demo.launch()
```

### Logs

```shell
Traceback (most recent call last):
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/gradio/routes.py", line 414, in run_predict
    output = await app.get_blocks().process_api(
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/gradio/blocks.py", line 1321, in process_api
    inputs = self.preprocess_data(fn_index, inputs, state)
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/gradio/blocks.py", line 1171, in preprocess_data
    processed_input.append(block.preprocess(inputs[i]))
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/gradio/components.py", line 1361, in preprocess
    return self.choices.index(x)
ValueError: 'Z' is not in list
```
```


### System Info

```shell
gradio==3.31.0
gradio_client==0.2.5
```

### Severity

blocking all usage of gradio

**Comment by pseudotensor** [comment]
I was able to work-around by keeping a state of a dictionary and keeping track of everything abou radio buttons myself instead of relying upon index.  So then I went back to 'value' and never rely upon the radio button to tell me what choices are present, only what was selected.

**Comment by freddyaboulton** [comment]
Hi @pseudotensor ! Yea this is an own bug. Radio indexing only works with the set of values used when the radio is first initiated. Your workaround is the correct thing to do while we fix it. I think there’s an existing issue for this - will close if I can find it!

**Comment by abidlabs** [comment]
Yup this is the same issue as https://github.com/gradio-app/gradio/issues/1566, which we will tackle soon

**Comment by pseudotensor** [comment]
is this related? https://github.com/gradio-app/gradio/discussions/2848

**Comment by abidlabs** [comment]
Ohh that's not a good idea. It'll change it for all users of your Gradio app, so if multiple people are using the app at the same time, there could be issues. We'll work on a proper fix, but it's a bit tricky to get it working. Might need an internal breaking change, so we'll likely ship it as part of 4.0

**Comment by myke11j** [comment]
@abidlabs has there been any fix on this? I'm using gradio=5.6.0, and facing the same issue of not being to update value in a dropdown because it was already in the options when it was created during the interface load.

```
gradio.exceptions.Error: "Value: <value> is not in the list of choices: <options_list>"
```

**Comment by abidlabs** [comment]
@myke11j are you using the latest version of Gradio? If so can you please create a new issue with a minimal repro? This issue should have been fixed
